### PR TITLE
Refactor Range Search model to include boost variant 

### PR DESCRIPTION
**Pull request by dineshraj01**

### CODE REVIEW

File: src/mlpack/methods/range_search/rs_model.hpp [hunk 1]
```diff
@@ -19,12 +19,212 @@
...
+  void operator()(RSTypeT<tree::KDTree>* rs) const;
+
+  //! Train on the given RSType specialized for BallTrees.
+  void operator()(RSTypeT<tree::BallTree>* rs) const;
+
+  //! Train specialized for octrees.
+  void operator()(RSTypeT<tree::Octree>* rs) const;
+
+  //! Construct the TrainVisitor object with the given reference set, leafSize
+  //! for BinarySpaceTrees, and tau and rho for spill trees.
```
Comment (rcurtin) [inline src/mlpack/methods/range_search/rs_model.hpp]: Can you fix the comment here please?  tau and rho and spill trees aren't part of RangeSearch.

File: src/mlpack/methods/range_search/rs_model.hpp [hunk 2]
```diff
@@ -55,45 +255,26 @@ class RSModel
...
+                 RSType<tree::BallTree> *,
+                 RSType<tree::XTree> *,
+                 RSType<tree::HilbertRTree> *,
+                 RSType<tree::RPlusTree> *,
+                 RSType<tree::RPlusPlusTree> *,
+                 RSType<tree::VPTree> *,
+                 RSType<tree::RPTree> *,
+                 RSType<tree::MaxRPTree> *,
+                 RSType<tree::UBTree> *,
+                 RSType<tree::Octree> *> rSearch;
```
Comment (rcurtin) [inline src/mlpack/methods/range_search/rs_model.hpp]: This is a minor style comment---no space is needed between the type and the pointer, so `RSType<tree::UBTree>*` not `RSType<tree::UBTree> *` please.

File: src/mlpack/methods/range_search/rs_model_impl.hpp [hunk 3]
```diff
@@ -18,6 +18,216 @@
...
+  throw std::runtime_error("no range search model initialized");
+}
+
+//! Save parameters for bichromatic range search.
+BiSearchVisitor::BiSearchVisitor(const arma::mat& querySet,
+                                                        const math::Range& range,
+                                                        std::vector<std::vector<size_t>>& neighbors,
+                                                        std::vector<std::vector<double>>& distances,
+                                                        const size_t leafSize
+                                                        ) :
```
Comment (rcurtin) [inline src/mlpack/methods/range_search/rs_model_impl.hpp]: Looks like something strange happened with the spacing here; be sure your editor is using spaces and not tab characters.  (I think we really need to get an automatic style checker running for PRs, then there will be fewer random comments from me! :))

File: src/mlpack/methods/range_search/rs_model_impl.hpp [hunk 4]
```diff
@@ -18,6 +18,216 @@
...
+       typename RSType::Tree* tree =
+                    new typename RSType::Tree(std::move(referenceSet),
+                    oldFromNewReferences, leafSize);
+       rs->Train(tree);
+ 
+     // Give the model ownership of the tree and the mappings.
+     rs->treeOwner = true;
+     rs->oldFromNewReferences = std::move(oldFromNewReferences);
+  }
+      }
```
Comment (rcurtin) [inline src/mlpack/methods/range_search/rs_model_impl.hpp]: Spacing is strange here too.

Comment (dineshraj01) [inline src/mlpack/methods/range_search/rs_model_impl.hpp]: oh thanks for pointing out...I'll fix the irregularities in coding style as soon as possible and also make sure not to repeat this thing in future.

**Comment by dineshraj01** [comment]
Hi, I have made some changes and fixed the wrong coding style used, kindly review.

**Comment by rcurtin** [comment]
It was easier to just fix the style than comment on it, so I made those changes in 6b2d239.  Thanks again for the contribution. :)

**Comment by dineshraj01** [comment]
👍 
and got my lesson :) thanks.

**Pull request merged**